libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1681,161 +1681,198 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
+
+		// Will be assigned to .type at the end (turning multi-elements into a tuple).
+		bool const allowDynamicTypes = m_evmVersion.supportsReturndata();
+		returnTypes = allowDynamicTypes
+			? _functionType->returnParameterTypes()
+			: _functionType->returnParameterTypesWithoutDynamicTypes();
 	}
 
-	if (functionType->kind() == FunctionType::Kind::BareStaticCall && !m_evmVersion.hasStaticCall())
-		m_errorReporter.typeError(_functionCall.location(), "\"staticcall\" is not supported by the VM version.");
+	if (returnTypes.size() == 1)
+		return returnTypes.front();
+
+	return make_shared<TupleType>(returnTypes);
+}
+
+void TypeChecker::typeCheckNonABIDecodeFunctionLike(
+	FunctionCall const& _functionCall,
+	FunctionTypePointer _functionType
+)
+{
+	// Actual function call or struct constructor call.
+
+	solAssert(!!_functionType, "");
+	solAssert(_functionType->kind() != FunctionType::Kind::ABIDecode, "");
+
+	bool const isPositionalCall = _functionCall.names().empty();
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+	vector<ASTPointer<ASTString>> const& argumentNames = _functionCall.names();
+
+	if (
+		_functionType->kind() == FunctionType::Kind::BareStaticCall &&
+		!m_evmVersion.hasStaticCall()
+	)
+		m_errorReporter.typeError(
+			_functionCall.location(),
+			"\"staticcall\" is not supported by the VM version."
+		);
 
 	if (auto functionName = dynamic_cast<Identifier const*>(&_functionCall.expression()))
 	{
-		if (functionName->name() == "sha3" && functionType->kind() == FunctionType::Kind::KECCAK256)
-			m_errorReporter.typeError(_functionCall.location(), "\"sha3\" has been deprecated in favour of \"keccak256\"");
-		else if (functionName->name() == "suicide" && functionType->kind() == FunctionType::Kind::Selfdestruct)
-			m_errorReporter.typeError(_functionCall.location(), "\"suicide\" has been deprecated in favour of \"selfdestruct\"");
+		if (
+			functionName->name() == "sha3" &&
+			_functionType->kind() == FunctionType::Kind::KECCAK256
+		)
+			m_errorReporter.typeError(
+				_functionCall.location(),
+				"\"sha3\" has been deprecated in favour of \"keccak256\""
+			);
+		else if (
+			functionName->name() == "suicide" &&
+			_functionType->kind() == FunctionType::Kind::Selfdestruct
+		)
+			m_errorReporter.typeError(
+				_functionCall.location(),
+				"\"suicide\" has been deprecated in favour of \"selfdestruct\""
+			);
 	}
-	if (!m_insideEmitStatement && functionType->kind() == FunctionType::Kind::Event)
-		m_errorReporter.typeError(_functionCall.location(), "Event invocations have to be prefixed by \"emit\".");
-
-	TypePointers parameterTypes = functionType->parameterTypes();
+	if (!m_insideEmitStatement && _functionType->kind() == FunctionType::Kind::Event)
+		m_errorReporter.typeError(
+			_functionCall.location(),
+			"Event invocations have to be prefixed by \"emit\"."
+		);
 
-	if (!functionType->padArguments())
+	if (!_functionType->padArguments())
```

Comment:
This is only true for the `abi.encodePacked` function.

Reply:
The if condition would be true for all eight of the following FunctionType::Kind values; it's not apparent to me that the other seven values would be filtered out in advance at this point.

    case Kind::BareCall:
    case Kind::BareCallCode:
    case Kind::BareDelegateCall:
    case Kind::BareStaticCall:
    case Kind::SHA256:
    case Kind::RIPEMD160:
    case Kind::KECCAK256:
    case Kind::ABIEncodePacked:

Reply:
You are right, I forgot about the other 7. The thing is, though, that all except `ABIEncodePacked` only have a single parameter of type `bytes memory` (we changed that recently) and rational number literals cannot be converted to `bytes memory`. So the check below actually only is meaningful for `ABIEncodePacked`.

Reply:
I've moved variadic and packed argument checks to a separate method for the various ABI encode* functions.